### PR TITLE
Replace rulesengine with kuiper in the compose-file

### PR DESCRIPTION
**Pull request by zhanghongtong**
Replace rulesengine with kuiper in the compose-file

### CODE REVIEW

File: releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty-arm64.yml [hunk 1]
```diff
@@ -226,15 +226,22 @@ services:
...
       edgex-network:
         aliases:
-          - edgex-support-rulesengine
+          - edgex-kuiper
+    environment:
+      EDGEX_SERVER: edgex-core-data
+      EDGEX_SERVICE_SERVER: http://edgex-core-data:48080
+      EDGEX_TOPIC: events
+      EDGEX_PROTOCOL: tcp
+      EDGEX_PORT: 5563
```
Comment (lenny-intel) [inline releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty-arm64.yml:244]: Port needs to be `5566` so you receive events from app-service-rules that may be doing some filtering.

Also need to add environment to app-service-rules to set the Publish topic to be `events` as it is currently blank. 
Added the following for app-service-rules environment:
```
Binding_PublishTopic: events
```

File: releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml [hunk 2]
```diff
@@ -224,15 +224,22 @@ services:
...
       edgex-network:
         aliases:
-          - edgex-support-rulesengine
+          - edgex-kuiper
+    environment:
+      EDGEX_SERVER: edgex-core-data
+      EDGEX_SERVICE_SERVER: http://edgex-core-data:48080
+      EDGEX_TOPIC: events
+      EDGEX_PROTOCOL: tcp
+      EDGEX_PORT: 5563
```
Comment (lenny-intel) [inline releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml:242]: Port needs to be `5566` so you receive events from app-service-rules that may be doing some filtering.

Also need to add environment to app-service-rules to set the Publish topic to be `events` as it is currently blank. 
Added the following for app-service-rules environment:
```
Binding_PublishTopic: events
```

File: releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml [hunk 3]
```diff
@@ -224,15 +224,22 @@ services:
...
+      - "20498:20498"
+    container_name: edgex-kuiper
+    hostname: edgex-kuiper
     networks:
       edgex-network:
         aliases:
-          - edgex-support-rulesengine
+          - edgex-kuiper
+    environment:
+      EDGEX_SERVER: edgex-core-data
```
Comment (lenny-intel) [inline releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml:238]: Do you also need to specify ZQM vs MQTT, or are you defaulting to ZMQ?

Comment (jinfahua) [reply releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml:238]: Currently

File: releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty-arm64.yml (hunk 1 above)
Comment (zhanghongtong) [reply releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty-arm64.yml]: Thanks for your advice, I have updated the code

File: releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml (hunk 2 above)
Comment (zhanghongtong) [reply releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml]: Thanks for your advice, I have updated the code

**Comment by jinfahua** [comment]
@lenny-intel 

With the latest docker composer file, it seems that some of docker instances cannot be started, and we cannot verify our updates.

```shell
$ docker ps
CONTAINER ID        IMAGE                                                                  COMMAND                  CREATED             STATUS              PORTS                                                                                              NAMES
be454da284d1        nexus3.edgexfoundry.org:10004/docker-device-virtual-go:master          "/device-virtual --p…"   7 minutes ago       Up 7 minutes        0.0.0.0:49990->49990/tcp                                                                           edgex-device-virtual
19595347fa67        nexus3.edgexfoundry.org:10004/docker-edgex-ui-go:master                "./edgex-ui-server"      7 minutes ago       Up 7 minutes        0.0.0.0:4000->4000/tcp                                                                             edgex-ui-go
6fdfeaec65b2        emqx/kuiper:0.3.0                                                      "/usr/bin/docker-ent…"   7 minutes ago       Up 7 minutes        0.0.0.0:9081->9081/tcp, 0.0.0.0:20498->20498/tcp                                                   edgex-kuiper
2f2e6c8b374f        nexus3.edgexfoundry.org:10004/docker-app-service-configurable:master   "/app-service-config…"   7 minutes ago       Up 7 minutes        48095/tcp, 0.0.0.0:48100->48100/tcp                                                                edgex-app-service-configurable-rules
1e342c4dd972        consul:1.3.1                                                           "docker-entrypoint.s…"   8 minutes ago       Up 8 minutes        0.0.0.0:8400->8400/tcp, 8300-8302/tcp, 8301-8302/udp, 8600/tcp, 8600/udp, 0.0.0.0:8500->8500/tcp   edgex-core-consul
4edb4370a008        redis:5.0.7-alpine                                                     "docker-entrypoint.s…"   8 minutes ago       Up 8 minutes        0.0.0.0:6379->6379/tcp                                                                             edgex-redis
723572c12f6d        nexus3.edgexfoundry.org:10004/docker-edgex-volume:master               "/bin/sh -c '/usr/bi…"   8 minutes ago       Up 8 minutes                                                                                                           edgex-files
```

**Comment by lenny-intel** [comment]
> With the latest docker composer file, it seems that some of docker instances cannot be started, and we cannot verify our updates.

Between run be sure to "clean" you docker volumes so old data isn't hanging around. I do this with the following commands.

```
docker-compose -f <file> down
docker rm $(docker ps -aq)  # removes stopped containers that have hold on volumes
docker volume prune # removes unused volumes
```
I ran your `docker-compose-nexus-mongo-no-secty.yml` and all services running, except device-virtual (port issue I have on my system). I can start device-virtual from portianer and it is sending data. Using Consul I switched app-service-rules to DEBUG logging and see the data flowing there, but the `kuiper` logs don't show any data flowing.

I also verified app-service-rules is configured properly.

```
Initializing Message Bus Trigger. Subscribing to topic: events on port 5563 , Publish Topic: events on port 5566
```

Should I expect the `kuiper` logs to show data being received?

**Comment by jinfahua** [comment]
Now we can bring up all of Dockers successfully (we cleaned our docker volumes, and removed all of docker images in local). Regarding to your below question,
> I ran your `docker-compose-nexus-mongo-no-secty.yml` and all services running, except device-virtual (port issue I have on my system). I can start device-virtual from portianer and it is sending data. Using Consul I switched app-service-rules to DEBUG logging and see the data flowing there, but the `kuiper` logs don't show any data flowing.

No, you should not expect data flowing because NO rules are deployed in Kuiper by default. To take a look at the data flow into Kuiper, you need to take below steps.
- Create a stream
  1. Enter Kuiper docker instance. Type below command in docker host.
  ```
  docker exec -it edgex-kuiper /bin/sh
  ```
  2. In Kuiper docker instance, create a demo stream with following command.
  ```
  bin/cli create stream demo'() WITH (FORMAT="JSON", TYPE="edgex")'
  ```
  3. You can verify see if the stream is created successfully or not. It it is created successfully, then ``demo`` will be print.
  ```
  # bin/cli show streams
  Connecting to 127.0.0.1:20498...
  demo
  ```
- Run query
  1. Type following command in Kuiper docker instance,
  ```
  bin/cli query
  ```
  2. Type SQL in Kuiper command line
  ```
  kuiper > SELECT * FROM demo
  ```
  3. If it works, then the data will be print into the console. 

In my testing, it seems that the data does not flow into Kuiper. I cannot see any data in the last step. Could you help to verify in your environment? If you make sure that the data are flowed into Kuiper, but console does not print anything, could you please help to post contents of log file (``log/stream.log`` in ) here? Thanks.

**Comment by jinfahua** [comment]
@lenny-intel I figured out the problem - the host for EdgeX message bus is not correct. The port is changed to ``5566``, while host is still point to ``edgex-core-data``, so the data cannot be flowed into Kuiper. I will have @zhanghongtong to update composer file to change to host `` edgex-app-service-configurable-rules``.  

Below is the result after changing to correct configurations.

```shell
# cat etc/sources/edgex.yaml
#Global Edgex configurations
default:
  serviceServer: http://edgex-core-data:48080
  topic: events
  port: 5566
  server: edgex-app-service-configurable-rules
  protocol: tcp
#  optional:
#    ClientId: client1
#    Username: user1
#    Password: password
#Override the global configurations
application_conf: #Conf_key
  protocol: tcp
  server: localhost
  port: 5571
  topic: application/kuiper 

# bin/cli query
Connecting to 127.0.0.1:20498...
kuiper > select * from demo;
Query was submit successfully.
kuiper > [{"bool":false}]
[{"int8":58}][{"int64":-4004014153667254086}][{"int16":14832}][{"int32":-2060755826}]
[{"uint64":5358792673154920656}]
[{"uint16":36832}][{"uint8":46}][{"uint32":1397099565}]
[{"bool":true}]
```

**Comment by lenny-intel** [comment]
Excellent! good catch!

BTW can the Stream & Query be create via REST rather than having to exec into the running container?

**Comment by lenny-intel** [comment]
> I figured out the problem - the host for EdgeX message bus is not correct. The port is changed to `5566`, while host is still point to `edgex-core-data`,

Excellent! good catch!

BTW, can the Stream & Query be created via REST rather than having to exec into the running container?

**Comment by jinfahua** [comment]
> BTW, can the Stream & Query be created via REST rather than having to exec into the running container?

Yes, ``streams & rules`` can be managed by both [REST](https://github.com/emqx/kuiper/blob/develop/docs/en_US/restapi/overview.md) and CLI. But query is a tool for debugging SQLs, and analysis result will be print to console directly, so query cannot be used through REST.

**Comment by jinfahua** [comment]
@lenny-intel We fixed the problem, and could you please help to review the PR again? Thanks.

File: releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml [hunk 4]
```diff
@@ -218,21 +218,29 @@ services:
...
+    container_name: edgex-kuiper
+    hostname: edgex-kuiper
     networks:
       edgex-network:
         aliases:
-          - edgex-support-rulesengine
+          - edgex-kuiper
+    environment:
+      EDGEX_SERVER: edgex-app-service-configurable-rules
+      EDGEX_SERVICE_SERVER: http://edgex-core-data:48080
```
Comment (lenny-intel) [inline releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml:240]: Why do you need this?

**Comment by lenny-intel** [comment]
@jinfahua, i have tested using you tutorial to create stream and a rule via rest and connected MQTT.fx up to same broker subscribing on the `results` topic. Working great!! Please rebase you PR so we can prepare to merge. FYI, @jpwhitemn , @rsdmike

I still don't see anything in the Kuiper log file once the stream and rules have been created. Do you have log levels that we can control from the environment variables or even a REST call? It would be nice to have DEBUG logging once the stream has been created as a verification that we are getting data to flowing into the stream. It would also be great to have DEBUG logging when a rule has been executed and if no rules created it would log a message to that fact.

File: releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml (hunk 4 above)
Comment (jinfahua) [reply releases/nightly-build/compose-files/docker-compose-nexus-redis-no-secty.yml:240]: - ``edgex-app-service-configurable-rules`` is used for listening data from application service.
- ``http://edgex-core-data:48080`` is used for getting data types of data that is flowed into the Kuiper, and Kuiper will try to convert the ``EdgeX data type`` to ``Kuiper data type``. Refer to this link for the [data type conversions](https://github.com/emqx/kuiper/blob/master/docs/en_US/rules/sources/edgex.md).

**Comment by jinfahua** [comment]
> @jinfahua, i have tested using you tutorial to create stream and a rule via rest and connected MQTT.fx up to same broker subscribing on the `results` topic. Working great!! Please rebase you PR so we can prepare to merge. FYI, @jpwhitemn , @rsdmike
> 
> I still don't see anything in the Kuiper log file once the stream and rules have been created. Do you have log levels that we can control from the environment variables or even a REST call? It would be nice to have DEBUG logging once the stream has been created as a verification that we are getting data to flowing into the stream. It would also be great to have DEBUG logging when a rule has been executed and if no rules created it would log a message to that fact.

@lenny-intel Thanks for your verification. 
We will update the docker image to support environment variables for controlling log levels at Kuiper 0.3.1 version. The version will be released at the end of this week or early next week, and it will be the version that ships with EdgeX Geneva.

**Comment by lenny-intel** [comment]
@jinfahua , decision in today's TSC meeting is to comment out the old rules engine block rather than replace it. Could you please make this change and also rebase. THX!

File: releases/nightly-build/compose-files/docker-compose-nexus-mongo.yml [hunk 5]
```diff
@@ -435,21 +435,29 @@ services:
...
       - data
 
   rulesengine:
-    image: nexus3.edgexfoundry.org:10004/docker-support-rulesengine:master
+    image: emqx/kuiper:0.3.0
     ports:
-      - "48075:48075"
-    container_name: edgex-support-rulesengine
-    hostname: edgex-support-rulesengine
+      - "9081:9081"
```
Comment (lenny-intel) [inline releases/nightly-build/compose-files/docker-compose-nexus-mongo.yml:447]: Anyway to configure Kuiper to use our standard `48075` rules engine port? Our proxy setup initializes routes to rules engine via this port.

**Comment by jinfahua** [comment]
@lenny-intel Following changes are made, could you please review? Thanks.
- Old rules engine block rather than replace it, and rebased PR.
- Use 48075 instead of 9081
- Log are print console (``docker logs edgex-kuiper``)
- Upgraded Kuiper Docker image version from 0.3.0 to 0.3.1

**Comment by lenny-intel** [comment]
> @lenny-intel Following changes are made, could you please review? Thanks.

@jinfahua , this is great. I will test out these changes today before our meeting. THX!

**Comment by lenny-intel** [comment]
@jinfahua, Please add the following override to the global environment section in each compose file
```
Clients_RulesEngine_Host: edgex-kuiper
```

Also I am seeing the following errors in the rules engine log file:
```
time="2020-04-16T17:45:57Z" level=warning msg="fail to get value for Float32: strconv.ParseFloat: parsing \"fwtOaw==\": invalid syntax" file="edgex_source.go:127" rule=rule3

time="2020-04-16T17:45:57Z" level=warning msg="got an empty result, ignored" file="edgex_source.go:159" rule=rule3

time="2020-04-16T17:46:03Z" level=warning msg="fail to get value for Uint64: strconv.Atoi: parsing \"10796529505058023104\": value out of range" file="edgex_source.go:127" rule=rule3

time="2020-04-16T17:46:03Z" level=warning msg="got an empty result, ignored" file="edgex_source.go:159" rule=rule3

time="2020-04-16T17:46:04Z" level=warning msg="payload {\"device\":\"Random-Binary-Device\",\"origin\":1587059163516497585,\"readings\":[{\"origin\":1587059162980482977,\"device\":\"Random-Binary-Device\",\"name\":\"Binary\",\"binaryValue\":\"lAF4AMahmk5JdQ74cqaXuI5a7Czkp1q9 unmarshal fail: media type must be specified for binary values" file="edgex_source.go:118" rule=rule3
```

I am not seeing any other readings being logged or push to MQTT broker.

BTW, I am using you demo stream and rule.

**Comment by zhanghongtong** [comment]
Hi @lenny-intel , I have updated the code

**Pull request merged**